### PR TITLE
nixos/acme: add webroots to ReadWritePaths

### LINKED ISSUE #139310: ACME's systemd hardening conf make cert renewal fails for webroots outside of `/var/lib/acme`
### Describe the bug
Since 7a10478ea7b992ffa1f19f389e53df0fe2aa936d, all `/var` except `/var/lib/acme` gets mounted in a read-only fashion.

This behaviour will break the existing acme deployments having a `webroot` set outside of `/var/lib/acme`.

In my personal case, I have the webroot served by my main nginx. Meaning:

```
  security.acme = {
    email = "$redactedemail";
    acceptTerms = true;
    certs = {
      "hardin.alternativebit.fr" = {
        webroot = "/var/www/hardin.alternativebit.fr";
        email = "$redactedemail";
        group = "nginx";
      };
    };
  };
```

After 7a10478ea7b992ffa1f19f389e53df0fe2aa936d, the cert renewal fails with:

```
hardin systemd[1]: Starting Renew ACME certificate for hardin.alternativebit.fr...
hardin acme-hardin.alternativebit.fr-start[756450]: + mkdir -p /var/www/hardin.alternativebit.fr/.well-known/acme-challenge
hardin acme-hardin.alternativebit.fr-start[756450]: + chgrp nginx /var/www/hardin.alternativebit.fr/.well-known/acme-challe>
hardin acme-hardin.alternativebit.fr-start[756452]: chgrp: changing group of '/var/www/hardin.alternativebit.fr/.well-known>
hardin acme-hardin.alternativebit.fr-start[756453]: + echo 'Please ensure /var/www/hardin.alternativebit.fr/.well-known/acm>
hardin acme-hardin.alternativebit.fr-start[756453]: Please ensure /var/www/hardin.alternativebit.fr/.well-known/acme-challe>
hardin acme-hardin.alternativebit.fr-start[756453]: + exit 1
hardin systemd[1]: acme-hardin.alternativebit.fr.service: Main process exited, code=exited, status=1/FAILURE
hardin systemd[1]: acme-hardin.alternativebit.fr.service: Failed with result 'exit-code'.
hardin systemd[1]: Failed to start Renew ACME certificate for hardin.alternativebit.fr.
```

The fix I personally went with was to add my webroot to the `readwritepath` by overriding the service config.

```
  systemd.services."acme-hardin.alternativebit.fr".serviceConfig = {
    ReadWritePaths = [
      "/var/www/hardin.alternativebit.fr"
    ];
  };
```

That's however not a proper fix for NixPkgs. As it is, we'll silently break some already existing deployments. While that's something acceptable for nixos-unstable, I feel like we have to adopt a better migration path **before** the next stable release.

I have three options in mind:

1. We could revert the `ProtectSystem = "strict"` hardening option to `ProtectSystem = "full";`.
1. We could add the webroots to the `ReadWritePaths` config. Could this backfire in some way?
1. We could trigger a assertion error warning the user that some of their webroots are not in the `ReadWritePath`s and instruct them about how to fix the issue. 

I did not manage do decide which of these solutions is the best. What do you think?

### Notify maintainers

cc @m1cr0man  @mweinelt @NixOS/acme

### PR DESCRIPTION
**Pull request by picnoir**
Since 7a10478ea7b992ffa1f19f389e53df0fe2aa936d, all /var except
/var/lib/acme gets mounted in a read-only fashion. This behavior
breaks the existing acme deployments having a webroot set outside of
/var/lib/acme.

Collecting the webroots and adding them to the paths read/write
mounted to the systemd service runtime tree.

Fixes #139310

Manually tested on my personal deployment.

### CODE REVIEW

**Comment by NinjaTrappeur** [comment]
@GrahamcOfBorg test acme

**Comment by m1cr0man** [comment]
It won't be quite as easy as appending to the commonServiceConfig. Ideally:
- we dedupe webroots in the list, and
- we only add webroots relevant to the respective services for a given cert.

To solve the second point all you really have to do is add the webroots only in serviceConfig around line 243

**Comment by NinjaTrappeur** [comment]
@GrahamcOfBorg test acme

**Comment by NinjaTrappeur** [comment]
Mmmh, this new patch does not solve the issue anymore on my test config. 

Don't merge that PR yet, I'll have a closer look tonight.

**Comment by NinjaTrappeur** [comment]
Alright, it seems to work both on my personal deployment and in the NixOS test.

PTAL

@GrahamcOfBorg test acme

**Comment by mweinelt** [comment]
The commit message should start with `nixos/acme:`.

**Comment by NinjaTrappeur** [comment]
@GrahamcOfBorg test acme

**Comment by NinjaTrappeur** [comment]
The nixos test isn't failing on my computer. I'm confused, I don't understand what's happening.

I think that's about as far as I'll push this PR. Feel free to push/rebase on my branch to fix this issue.

**Comment by NinjaTrappeur** [comment]
Okay, I finally found what I was doing wrong here. I was injecting some `null` values to the ReadWritePaths list by mistake.

Systemd was then unable to parse the whole ReadWritePaths directive. And, as usual, was silently ignoring it instead of throwing an error.

Filtering out the `null` webroots fixes the issue.

@GrahamcOfBorg test acme

File: nixos/modules/security/acme.nix
Comment (mweinelt) [inline nixos/modules/security/acme.nix:246]: I wonder: Does the selfsign service really need access to the webroot? What for?

Comment (NinjaTrappeur) [reply nixos/modules/security/acme.nix]: Indeed, looking at the script, I don't think we need a write access for a selfsigned renewal.

> To solve the second point all you really have to do is add the webroots only in serviceConfig around line 243

^ line 243 being the self-signed certificate service, it lead me to think I was missing something.

That being said, we might both missing something here :smile: 

@m1cr0man did you mistakenly pointed to the self-signed renewal part instead of the acme part or are we actually missing something?

**Review by m1cr0man — APPROVED** [review]
Nice job, the webroots resolver is really comprehensive

**Pull request merged**
